File: app/src/main/java/io/github/sithengineer/motoqueiro/data/DataManager.java
package io.github.sithengineer.motoqueiro.data;

import io.github.sithengineer.motoqueiro.app.Preferences;
import io.github.sithengineer.motoqueiro.hardware.Accelerometer;
import io.github.sithengineer.motoqueiro.hardware.Gravity;
import io.github.sithengineer.motoqueiro.hardware.Gyroscope;
import io.github.sithengineer.motoqueiro.hardware.MiBandService;
import io.github.sithengineer.motoqueiro.hardware.capture.LatLng;
import io.github.sithengineer.motoqueiro.hardware.capture.MiBandData;
import io.github.sithengineer.motoqueiro.hardware.capture.RelativeCoordinates;
import io.github.sithengineer.motoqueiro.hardware.gps.Gps;
import rx.Observable;
import rx.Single;
import timber.log.Timber;

public class DataManager {

  private final Accelerometer accelerometer;
  private final Gyroscope gyroscope;
  private final Gravity gravity;
  private final Gps gps;
  private final MiBandService miBand;
  private final RideRepository rideRepo;
  private final Preferences preferences;

  public DataManager(Accelerometer accelerometer, Gyroscope gyroscope, Gravity gravity, Gps gps,
      MiBandService miBand, RideRepository rideRepo, Preferences preferences) {
    this.accelerometer = accelerometer;
    this.gyroscope = gyroscope;
    this.gravity = gravity;
    this.gps = gps;
    this.miBand = miBand;
    this.rideRepo = rideRepo;
    this.preferences = preferences;
  }

  public Observable<Void> gatherData() {
    // gps data generator observable
    Observable<Void> generateGpsObservable =
        gps.listen().flatMapSingle(pos -> handleGpsPositionCapture(pos)).map(__ -> null);

    // accelerometer sensor data generator observable
    Observable<Void> generateAccelObservable = accelerometer.listen()
        .flatMapSingle(accelData -> handleAccelerometerCapture(accelData))
        .onErrorResumeNext(err -> {
          Timber.e(err);
          return Observable.empty();
        })
        .map(__ -> null);

    // gyroscope sensor data generator observable
    Observable<Void> generateGravityObservable = gravity.listen()
        .flatMapSingle(gravityData -> handleGravityCapture(gravityData))
        .onErrorResumeNext(err -> {
          Timber.e(err);
          return Observable.empty();
        })
        .map(__ -> null);

    // gyroscope sensor data generator observable
    Observable<Void> generateGyroObservable = gyroscope.listen()
        .flatMapSingle(gyroData -> handleGyroscopeCapture(gyroData))
        .onErrorResumeNext(err -> {
          Timber.e(err);
          return Observable.empty();
        })
        .map(__ -> null);

    // miband heart rate sensor data generator observable
    Observable<Void> generateHeartRateObservable = miBand.listen()
        .flatMapSingle(heartData -> handleMiBandCapture(heartData))
        .onErrorResumeNext(err -> {
          Timber.e(err);
          return Observable.empty();
        })
        .map(__ -> null);

    return Observable.merge(generateGpsObservable, generateAccelObservable, generateGyroObservable,
        generateGravityObservable, generateHeartRateObservable);
  }

  private String getRideId() {
    return preferences.getRideId();
  }

  private Single<Long> handleMiBandCapture(MiBandData bandData) {
    return rideRepo.saveHeartRate(getRideId(), bandData.getHeartRateBpm());
  }

  private Single<Long> handleGpsPositionCapture(LatLng capture) {
    return rideRepo.saveGpsCoordinate(getRideId(), capture.getLat(), capture.getLng());
  }

  private Single<Long> handleAccelerometerCapture(RelativeCoordinates capture) {
    return rideRepo.saveAccelerometerCapture(getRideId(), capture.getXx(), capture.getYy(),
        capture.getZz());
  }

  private Single<Long> handleGravityCapture(RelativeCoordinates capture) {
    return rideRepo.saveGravityCapture(getRideId(), capture.getXx(), capture.getYy(),
        capture.getZz());
  }

  private Single<Long> handleGyroscopeCapture(RelativeCoordinates capture) {
    return rideRepo.saveGyroscopeCapture(getRideId(), capture.getXx(), capture.getYy(),
        capture.getZz());
  }
}

File: app/src/main/java/io/github/sithengineer/motoqueiro/data/DataManager.kt
package io.github.sithengineer.motoqueiro.data

import io.github.sithengineer.motoqueiro.app.Preferences
import io.github.sithengineer.motoqueiro.exception.EmptyRideId
import io.github.sithengineer.motoqueiro.hardware.Accelerometer
import io.github.sithengineer.motoqueiro.hardware.Gravity
import io.github.sithengineer.motoqueiro.hardware.Gyroscope
import io.github.sithengineer.motoqueiro.hardware.MiBandService
import io.github.sithengineer.motoqueiro.hardware.capture.LatLng
import io.github.sithengineer.motoqueiro.hardware.capture.MiBandData
import io.github.sithengineer.motoqueiro.hardware.capture.RelativeCoordinates
import io.github.sithengineer.motoqueiro.hardware.gps.Gps
import io.reactivex.Completable

class DataManager(private val accelerometer: Accelerometer, private val gyroscope: Gyroscope,
    private val gravity: Gravity, private val gps: Gps,
    private val miBand: MiBandService, private val rideRepo: RideRepository,
    private val preferences: Preferences) {

  private val rideId: Long?
    get() = preferences.rideId

  fun gatherData(): Completable {
    // gps data generator observable
    val generateGpsObservable = gps.listen().flatMapCompletable { pos ->
      handleGpsPositionCapture(pos)
    }

    // accelerometer sensor data generator observable
    val generateAccelObservable = accelerometer.listen()
        .flatMapCompletable { accelData -> handleAccelerometerCapture(accelData) }

    // gyroscope sensor data generator observable
    val generateGravityObservable = gravity.listen()
        .flatMapCompletable { gravityData -> handleGravityCapture(gravityData) }

    // gyroscope sensor data generator observable
    val generateGyroObservable = gyroscope.listen()
        .flatMapCompletable { gyroData -> handleGyroscopeCapture(gyroData) }

    // miband heart rate sensor data generator observable
    val generateHeartRateObservable = miBand.listen()
        .flatMapCompletable { heartData -> handleMiBandCapture(heartData) }

    return Completable.merge(
        arrayListOf(
            generateGpsObservable,
            generateAccelObservable,
            generateGyroObservable,
            generateGravityObservable,
            generateHeartRateObservable
        )
    )
  }

  private fun handleMiBandCapture(bandData: MiBandData): Completable {
    val currentRideId = rideId
    return if (currentRideId != null) {
      rideRepo.saveHeartRate(currentRideId, bandData.heartRateBpm)
    } else {
      Completable.error(EmptyRideId())
    }
  }

  private fun handleGpsPositionCapture(capture: LatLng): Completable {
    val currentRideId = rideId
    return if (currentRideId != null) {
      rideRepo.saveGpsCoordinate(currentRideId, capture.lat, capture.lng)
    } else {
      Completable.error(EmptyRideId())
    }
  }

  private fun handleAccelerometerCapture(capture: RelativeCoordinates): Completable {
    val currentRideId = rideId
    return if (currentRideId != null) {
      rideRepo.saveAccelerometerCapture(currentRideId, capture.xx, capture.yy,
          capture.zz)
    } else {
      Completable.error(EmptyRideId())
    }
  }

  private fun handleGravityCapture(capture: RelativeCoordinates): Completable {
    val currentRideId = rideId
    return if (currentRideId != null) {
      rideRepo.saveGravityCapture(currentRideId, capture.xx, capture.yy,
          capture.zz)
    } else {
      Completable.error(EmptyRideId())
    }
  }

  private fun handleGyroscopeCapture(capture: RelativeCoordinates): Completable {
    val currentRideId = rideId
    return if (currentRideId != null) {
      rideRepo.saveGyroscopeCapture(currentRideId, capture.xx, capture.yy,
          capture.zz)
    } else {
      Completable.error(EmptyRideId())
    }
  }
}
